### PR TITLE
MIDI-103: Gradio app

### PR DESCRIPTION
**Pull request by WojciechMat**
### App

You can run the app with

```python
python -m dashboard.gradio.main
```
You need to upload a midi file to the app and it will create an audio with predicted velocities or dstart.

I omit creating MyTokenizedDataset as tgt is not needed if we do not wish to count loss nor distance.

```mermaid
flowchart TD
    A(.mid) -->|upload| B(gradio.File)
    B --> C(MidiPiece)
    C -->|process_piece| D("list[list[token_id]]")
    D --> |model| E(predictions)
    C --> F(Predicted Piece)
    E --> F
```
### Default models
You can download default checkpoints with:
```shell
python checkpoints/download_models.py
```

### CODE REVIEW

**Comment by WojciechMat** [comment]
### New model just dropped :fire: 
You can download it [here](https://huggingface.co/wmatejuk/midi-velocity-transformer).
It was trained on giant-midi-sustain for one epoch but still manages to produce satisfactory results :)

**Comment by WojciechMat** [comment]
- [x] Enable uploading midi files
- [x] Predict velocity
- [x] Predict dstart
- [x] Show pianorolls
- [x] ? Make the video background a pianoroll ? (will research)
- [x] ? Show quantized midi piece ?
- [x] Add text

File: app/main.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import torch
+import gradio as gr
```
Comment (roszcz) [inline app/main.py]:
```sh
echo "gradio==3.47.1" >> requirements.txt
```

File: app/main.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import fortepyan as ff
+from omegaconf import OmegaConf
+from fortepyan.audio import render as render_audio
+
+from app.tools import load_model, predict_velocity
+
+
+def run_app(target: str, midi_file):
+    checkpoint = torch.load("checkpoints/velocity/midi-transformer-2023-10-03-08-34.pt", map_location="cpu")
```
Comment (roszcz) [inline app/main.py]: 1. Can you make it possible to select one of the checkpoints present in the folder?
2. If you have a well performing model in HF, can you add code that will download it?

File: app/main.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    demo = gr.Interface(
+        run_app,
+        inputs=[
+            gr.components.Radio(["velocity"]),
+            gr.File(),
+        ],
+        outputs=[gr.Video()],
+    )
+
+    demo.launch()
```
Comment (roszcz) [inline app/main.py]:
```suggestion
    demo.launch(server_name="0.0.0.0")
```

This will allow access through the network - I have a multi-machine setup, and access over wifi is important to me :radio:

File: app/tools.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+from utils import vocab_sizes, greedy_decode
+from data.dataset import quantized_piece_to_records
+from data.tokenizer import VelocityEncoder, QuantizedMidiEncoder
+
+
+def encode_record(sequence: dict, encoder: QuantizedMidiEncoder) -> torch.Tensor:
+    token_ids = encoder.encode(sequence)
+    token_ids = [encoder.token_to_id["<CLS>"]] + token_ids
+    return torch.tensor(token_ids, dtype=torch.int64)
+
```
Comment (roszcz) [inline app/tools.py]: I think this should be a method of the encoder itself.

File: app/tools.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    src_encoder = QuantizedMidiEncoder(train_cfg.dataset.quantization)
+
+    # pre-process the piece ...
+    qpiece = quantizer.inject_quantization_features(piece)
+    sequences = quantized_piece_to_records(
+        piece=qpiece,
+        sequence_len=train_cfg.dataset.sequence_len,
+        sequence_step=train_cfg.dataset.sequence_len,
+    )
+    sequence_tokens = [encode_record(sequence, src_encoder) for sequence in sequences]
```
Comment (roszcz) [inline app/tools.py]:
```suggestion
    sequence_tokens = [src_encoder.encode_record(sequence) for sequence in sequences]
```
see the comment about `encode_record` :arrow_up:

File: app/tools.py (hunk 4 above)
Comment (WojciechMat) [reply app/tools.py]: The idea was to not append CLS token during encoding inside the encoder.

Comment (roszcz) [reply app/tools.py]: My point is that from object oriented point of view, I'd prefer not to do:

```py
# No
def some_method(payload, some_object):
    return some_object.some_method(payload)
```

But go in this direction:

```py
# Yes
class SomeObject:
    def some_method(self, payload):
        ...
```

Also, since you are only calling this function in one place, you can also write the two lines of code explicitly there, and avoid obfuscation :)

**Comment by WojciechMat** [comment]
### New dstart model
There is new model at [huggingfacce](https://huggingface.co/wmatejuk/midi-dstart-transformer)

I have also added a script to download default models:
```shell
python checkpoints/download_models.py
```
You can now also run dstart prediction on the app.

**Comment by WojciechMat** [comment]
Pianorolls are now displayed as well.

File: checkpoints/download_models.py
Comment (roszcz) [inline checkpoints/download_models.py:1]: We shouldn't have directories with mixed content type (python files and checkpoints). I think we should put all dashboard code in a single folder module, and maybe have a structure like this:

```sh
├── dashboard
│   ├── download_models.py
│   ├── gradio
│   │   ├── app.py
│   ├── streamlit
│   │   ├── dstart
│   │   ├── velocity
```

File: checkpoints/download_models.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+from huggingface_hub import hf_hub_download
+
+FILENAME_VELOCITY = "midi-transformer-2023-10-03-08-34.pt"
+FILENAME_DSTART = "midi-transformer-2023-10-06-11-55.pt"
+hf_hub_download(repo_id="wmatejuk/midi-velocity-transformer", filename=FILENAME_VELOCITY, local_dir="velocity")
+hf_hub_download(repo_id="wmatejuk/midi-dstart-transformer", filename=FILENAME_DSTART, local_dir="dstart")
```
Comment (roszcz) [inline checkpoints/download_models.py]:
```suggestion
hf_hub_download(
    repo_id="wmatejuk/midi-velocity-transformer",
    filename=FILENAME_VELOCITY,
    local_dir="checkpoints/velocity",
    local_dir_use_symlinks=False,
)
hf_hub_download(
    repo_id="wmatejuk/midi-dstart-transformer", 
    filename=FILENAME_DSTART, 
    local_dir="checkpoints/dstart",
    local_dir_use_symlinks=False,
)
```

Right now this creates new `velocity` and `dstart` folders in the root folder of ther repo, and that's not where the dashboards expect them to be. 

Also, without setting `local_dir_use_symlinks=False` the file is downloaded to the HF cache and in the folder we only get a link which I think is less convenient.

**Review by roszcz — COMMENTED** [review]
This works great, but I have some change requests:

- [x] apply both `velocity` and `dstart` models to a each uploaded MIDI file
- [x] make to MIDI after changes downloadable
- [ ] display piano rolls for intermediary steps, i.e. input, after dstart,after velocity, final

**Comment by WojciechMat** [comment]
Pianorolls are displayed as a backgroud of the video:
image

**Comment by roszcz** [comment]
I encountered an edge case with this file:

[roszcz_maestro-v1-sustain-test-108-100-240-0.42Random_Mask.mid.tar.gz](https://github.com/Nospoko/midi-translation/files/12843228/roszcz_maestro-v1-sustain-test-108-100-240-0.42Random_Mask.mid.tar.gz)

Looks like the `predicted_piece` is empty

```
  File "/home/hagrid/workson/midi-translation/dashboard/gradio/main.py", line 38, in run_predict_app
    vid = make_pianoroll_video(predicted_piece)
  File "/home/hagrid/workson/midi-translation/dashboard/gradio/tools.py", line 162, in make_pianoroll_video
    ff.view.draw_pianoroll_with_velocities(piece)
  File "/home/hagrid/.virtualenvs/midi-translation/lib/python3.10/site-packages/fortepyan/view/pianoroll/main.py", line 30, in draw_pianoroll_with_velocities
    piano_roll = PianoRoll(piece, time_end=time_end)
  File "<string>", line 9, in __init__
  File "/home/hagrid/.virtualenvs/midi-translation/lib/python3.10/site-packages/fortepyan/view/pianoroll/structures.py", line 26, in __post_init__
    self._build_image()
  File "/home/hagrid/.virtualenvs/midi-translation/lib/python3.10/site-packages/fortepyan/view/pianoroll/structures.py", line 48, in _build_image
    n_time_steps = self.RESOLUTION * int(np.ceil(self.duration))
ValueError: cannot convert float NaN to integer
```

**Comment by WojciechMat** [comment]
> I encountered an edge case with this file:
> 
> [roszcz_maestro-v1-sustain-test-108-100-240-0.42Random_Mask.mid.tar.gz](https://github.com/Nospoko/midi-translation/files/12843228/roszcz_maestro-v1-sustain-test-108-100-240-0.42Random_Mask.mid.tar.gz)
> 
> Looks like the `predicted_piece` is empty

This error occures when the midi piece is shorter than cfg.dataset.sequence_len. `quantized_piece_to_records` will treat it like the end of a sequence and resing from appending it.

I think the solution is:
```python
# pre-process the piece ...
    qpiece = quantizer.inject_quantization_features(piece)
    
    sequence_len = min(train_cfg.dataset.sequence_len, len(qpiece.df))
    sequences = quantized_piece_to_records(
        piece=qpiece,
        sequence_len=sequence_len,
        sequence_step=sequence_len,
    )
```
Will implement this soon.

**Comment by roszcz** [comment]
Not sure if the last commit was supposed to resolve this completely, but the app still errors with file I shared :)

**Comment by WojciechMat** [comment]
> Not sure if the last commit was supposed to resolve this completely, but the app still errors with file I shared :)

Are you sure you are not reusing the previous mp4? I think you can try and run 
```
cd /tmp
rm *.mp4 *.wav *.mp3 
```
and run gradio app again.
I might change tmp path for gradio somehow to our tpm/dashboard/gradio.

**Comment by roszcz** [comment]
Thanks, it works after a restart. While testing some midi files, I noticed that the dstart model has a tendency to insert very long pauses occasionally, that do not make any musical sense:

image

Do you have any ideas where this may be coming from and what to do to prevent this?

**Comment by WojciechMat** [comment]
> Thanks, it works after a restart. While testing some midi files, I noticed that the dstart model has a tendency to insert very long pauses occasionally, that do not make any musical sense:
> 
> image
> Do you have any ideas where this may be coming from and what to do to prevent this?


The last two of 200 dstart bin edges look like this:
```
  - 2.400000
  - 10.451823
```

So it seems like every dstart longer than 2.4 was extended to ~6.5 or worse. So I guess we need more bins for longer durations so that the last bin will not be present so often.

**Review by roszcz — APPROVED** [review]
Looks good to me, let me know if there's anything else you want to add here, and if not let's merge this! 🚀

**Comment by WojciechMat** [comment]
Right now dataflow when predicting both dstart and velocity looks like this:
```mermaid
flowchart TD

    A(MidiPiece) -->|quantizer| B(Quantized Piece for velocity)
    A -->|quantizer| L(Quantized Piece for dstart)
    B --> |velocity model|C(velocity)
    L -->|dstart_model| D("dstart")
    C --> E(Predicted Piece)
    D --> E(Predicted Piece)
    A --> F(duration)
    F -->E(Predicted Piece)
```
Maybe we could also do something like this:
```mermaid
flowchart TD

    A(MidiPiece) -->|quantizer| B(Quantized Piece for velocity)
    B --> |velocity model|C(velocity)
    C --> E(Predicted Velocity Piece)
    A --> E
    E -->|quantizer|F(Quantized Piece for dstart) 
    F --> |dstart model|G(dstart)
    E --> H(duration)
    E --> L(velocity)
    H --> K(Predicted Piece)
    G --> K
    L --> K
```
So that this behaves as a one model predicting velocity and dstart sequentially.
I am training a model on 128 velocity bin edges so that it wouldbe able to take full advantage of velocity predictions.

What do you think?

**Comment by WojciechMat** [comment]
+ models have to be trained on the same quantization bins in order to work right now

**Comment by WojciechMat** [comment]
Sequential prediction from previous comment is now implemented, download_models also downloads a dstart model trained on 128 velocity bins to make it possible.

Requesting quick approval and I think we can merge then.

**Review by roszcz — APPROVED** [review]
